tensorflow-lite/2.9.1

talyz opened the pull request:
Add **tensorflow-lite/2.9.1** (replacing **2.8.0** from the prior PR) and **tensorflow-lite/2.6.2**.

This picks up where #9684 left off. I rebased it on master, got rid of some unnecessary commits, and rearranged and squashed the remaining ones to produce an easy-to-read history.

All packages build fine on Linux, Windows and MacOS in my tests.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

prince-chrismc commented on recipes/tensorflow-lite/all/CMakeLists.txt:
```diff
@@ -1,8 +1,44 @@
+endif()
+
+find_package(farmhash REQUIRED)
+add_library(farmhash INTERFACE IMPORTED)
+set_property(TARGET farmhash PROPERTY INTERFACE_LINK_LIBRARIES farmhash::farmhash)
```
Why not use alias? https://cmake.org/cmake/help/latest/command/add_library.html#alias-libraries

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +83,46 @@ def requirements(self):
             self.requires("fp16/cci.20210320")
 
+    def build_requirements(self):
+        self.tool_requires("cmake/[^3.16]")
```
https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-are-version-ranges-not-allowed

We suggest just setting it to the latest version when a newer one is required

talyz replied:
done!

prince-chrismc commented on recipes/tensorflow-lite/all/conandata.yml:
```diff
@@ -2,11 +2,13 @@ sources:
-    - patch_file: "patches/non_x86_support.patch"
+  "2.9.1":
+    - patch_file: "patches/msvc_fixes.patch"
       base_path: "src"
```
This looks out of place, https://github.com/conan-io/conan-center-index/blob/master/docs/reviewing.md#subfolder-properties 

***

UPDATE: dont change this it will go away with v2 migration

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +83,46 @@ def requirements(self):
+        # Resolve the conflict by removing the conflicting
+        # implementation for now.
+        if tools.Version(self.version) >= "2.9.1" and self.settings.os == "Windows" and self.options.shared:
+            os.remove(os.path.join(self.source_folder, self._source_subfolder, "tensorflow/lite/simple_memory_arena_debug_dump.cc"))
```
can this be moved to a patch? it's more visible

talyz replied:
done!

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +83,46 @@ def requirements(self):
+        if tools.Version(self.version) >= "2.9.1" and self.settings.os == "Windows" and self.options.shared:
+            os.remove(os.path.join(self.source_folder, self._source_subfolder, "tensorflow/lite/simple_memory_arena_debug_dump.cc"))
+
+        # All CMake targets will be provided by Conan through the wrapper, removing every occurrence of `find_package`
```
🤢 Upstream has "special" cmake modules..... https://github.com/tensorflow/tensorflow/blob/539438c52dca142c13f244aa0a084698a978bcda/tensorflow/lite/tools/cmake/modules/FindFlatbuffers.cmake

prince-chrismc reviewed (commented):
First and foremost thank you for making time to adda new version, lots of people will be very happy!

I see you opted for programmatic replace/patching instead of the patching method that was already in place... 

I am curious if there was a technical reason behind this change? Or what your motivation was?

It's generally preferred to use patches with https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields we now have better visibility about what is being changed and why.

talyz commented:
> I see you opted for programmatic replace/patching instead of the patching method that was already in place...
> 
> I am curious if there was a technical reason behind this change? Or what your motivation was?
> 
> It's generally preferred to use patches with https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields we now have better visibility about what is being changed and why.

I just pulled this change, along with all the other ones authored by @hardsetting, from #9684, so I can't really speak to the motivation. I've just cleaned up and rebased the changes. I'm guessing it's a matter of simplification? I don't really have an opinion either way - I can bring the patches back if that's preferable.

prince-chrismc commented:
My concern is given this project is active, we dont know what impact it could have in the future... so it's safer to make patches so we can review what's actually being changed to make sure there is no harmful side affects.

I personally prefer the patches... we dont need to dynamically inject values so I think it's a little overkill
I'll admit it's more work upfront but there a better quality control we can enforce.

***

Would you mind if I took a shot at updating this? I have a few hours waiting at a dealership... this looks like the right amount of challenge while I wait

Looking that upstream CMake I can help but wonder if there is a less intrusive way

https://github.com/tensorflow/tensorflow/blob/539438c52dca142c13f244aa0a084698a978bcda/tensorflow/lite/tools/cmake/modules/Findfarmhash.cmake#L19

Instead of deleting a bunch of files/lines.. could we just set populated to true?

talyz commented:
Not at all, go ahead!

hardsetting commented:
Thanks @talyz for reviving and pushing this!
Chiming in regarding the following question:
> I see you opted for programmatic replace/patching instead of the patching method that was already in place...
I am curious if there was a technical reason behind this change? Or what your motivation was?

I went for dynamic patching as it gave a chance to reuse the same patch for multiple versions (at least under the same major version e.g. 2.8.x).
You raise a good point about better quality control, so I'm not opposing going back to "static" patching

talyz commented:
I've created a follow-up to this in #12390, where I try to migrate as much as possible to Conan v2.

talyz commented:
Seems like the CI got stuck for some reason? I can't see any details - maybe it just needs to be triggered again?

prince-chrismc commented on recipes/tensorflow-lite/all/conandata.yml:
```diff
@@ -2,11 +2,21 @@ sources:
-  "2.6.0":
-    - patch_file: "patches/local_find_packages.patch"
+  "2.9.1":
+    - patch_file: "patches/msvc_fixes.patch"
```
can you please add the extra metadata so we can track these 🙏 

https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

talyz replied:
Hm, when I do, I get the following error:
>[HOOK - conan-center.py] pre_export(): ERROR: [CONANDATA.YML FORMAT (KB-H030)] Additional entries ['patch_description', 'patch_type'] not allowed in 'patches':'2.9.1' of conandata.yml (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H030)

prince-chrismc replied:
You might need to update your hooks 

https://github.com/conan-io/hooks/blob/cb2a584f66666fac92bc0602c3d43cb3b7ee8f9a/hooks/conan-center.py#L368

See https://github.com/conan-io/conan-center-index/blob/9c5c6665ee5a1800d1e4b05ff23456c1f88ccaf8/docs/how_to_add_packages.md#updating-conan-hooks-on-your-machine

talyz replied:
Ah, yep, that was it :)

talyz replied:
done!

prince-chrismc commented:
So I spent a good part of my day yesterday trying to try a different approach.

I discovered there are a bunch of vendored libraries that are not being provided by Conan, some are not available in CCI 🙃 I also tried to migrate to v2... The issue you describe I hit as well...  with the new generators, upstream has a BUILD file the prevents the generator from writing files.

Since this add versions and does not make it worse, I think it's good to merge if you can just address https://github.com/conan-io/conan-center-index/pull/11477#discussion_r950317453

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +86,39 @@ def requirements(self):
-    def build(self):
+        files.get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
         for patch in self.conan_data.get("patches", {}).get(self.version, []):
             tools.patch(**patch)
```
There's `files.apply_conandata_patch`

prince-chrismc commented:
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

talyz commented:
@prince-chrismc Oh, sorry about that! I'm used to that being the workflow to keep the history somewhat sane. How do you usually do that? Should I just add fix-ups and then apply them when the review is done? :)

uilianries commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +86,39 @@ def requirements(self):
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
+        files.get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
+        files.apply_conandata_patches(self)
```
Patch should be applied under build. Move to there again.

uilianries reviewed (commented):
- Update `required_conan_version = ">=1.43.0"` to 1.50.0 due conan.errors

prince-chrismc commented:
> I'm used to that being the workflow to keep the history somewhat sane. How do you usually do that? Should I just add fix-ups and then apply them when the review is done? :)

You can just push commits, the bot will do a squash merge so you do not need to worry

Obviously good commit message still help since we can read what changed in the PR 😄

prince-chrismc commented on recipes/tensorflow-lite/all/conandata.yml:
```diff
@@ -2,11 +2,29 @@ sources:
+  "2.9.1":
+    - patch_file: "patches/msvc_fixes.patch"
+      patch_description: "Fix Windows build: Apply fixes for compatibility with MSVC from master"
+      patch_type: "backport"
```
Since this is a backpor, its best to link so we can see which version it was added too

```suggestion
      patch_type: "backport"
      patch_source: "https://github.com/tensorflow/tensorflow/commit/79be9efa5190a4a9c828369ba358e6f9e59fbbe5"
```

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +86,40 @@ def requirements(self):
         if self.options.with_xnnpack:
-            self.requires("xnnpack/cci.20211210")
+            self.requires("xnnpack/cci.20220621")
+            self.requires("pthreadpool/cci.20210218")
```
there's no specific call out for this and it's indirect https://conan.io/center/xnnpack?tab=dependencies

```suggestion
```

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +86,40 @@ def requirements(self):
+
     def validate(self):
         if self.settings.compiler.get_safe("cppstd"):
             tools.check_min_cppstd(self, 14)
```
Not sure it's worth it but in 2.9 they bumped it to 17... though everything is passing 🙈

prince-chrismc commented on recipes/tensorflow-lite/all/patches/dependencies_2_9.patch:
```diff
@@ -0,0 +1,87 @@
+     FILTER ".*(_test|_tester)\\.(cc|h)"
+   )
+   list(APPEND TFLITE_TARGET_DEPENDENCIES
+     XNNPACK
++    pthreadpool
```
Now I see why you called out this dep, I fixed it with

```suggestion
-     XNNPACK
+    xnnpack::xnnpack
```

talyz replied:
Oh, that does indeed solve the issue. How does that work? Why is `XNNPACK` different from `xnnpack::xnnpack`?

prince-chrismc replied:
There's no upstream CMake targets for that project, so tensorflow-lite invented it's own... only the default one for Conan is created https://conan.io/center/xnnpack?tab=useit

prince-chrismc reviewed (commented):
This is looking super good....

I am nit picking a little bit with this one, but it's bad upstream CMake not your work!

My only "must" is the `patch_source` comment so we can keep track

prince-chrismc commented:
This is a really hard package to get out so amazing work ❤️

prince-chrismc commented:
Let's get this merged!

uilianries commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +86,40 @@ def requirements(self):
+        # simple_memory_arena_debug_dump.obj : error LNK2005: "void __cdecl tflite::DumpArenaInfo(...) already defined in simple_memory_arena.obj
+        # Resolve the conflict by removing the conflicting
+        # implementation for now.
+        if Version(self.version) >= "2.9.1" and self.settings.os == "Windows" and self.options.shared:
```
Is there a related issue on the upstream? Also, it could be managed directly from cmake, excluding that file from the target.

prince-chrismc replied:
The CMake build scripts use GLOB expressions it's not called out anywhere https://github.com/tensorflow/tensorflow/search?q=simple_memory_arena_debug_dump,

in Bazel this is a separate lib https://github.com/tensorflow/tensorflow/blob/1a17caa08363e88204a285f6a8678c3b96e96e0e/tensorflow/lite/BUILD#L330

talyz replied:
I've changed to patching `CMakeLists.txt` to exclude the file there instead.

talyz commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -59,6 +62,8 @@ def export_sources(self):
         self.copy("CMakeLists.txt")
         for patch in self.conan_data.get("patches", {}).get(self.version, []):
             self.copy(patch["patch_file"])
+        if Version(self.version) >= "2.9.1":
+            self.copy("patches/remove_simple_memory_arena_debug_dump.patch")
```
Is that possible when the criteria for applying it is as niche as it is? Look at the code that applies the patch.

talyz replied:
Resolved by patching CMakeLists.txt instead.

prince-chrismc commented on recipes/tensorflow-lite/all/conanfile.py:
```diff
@@ -83,40 +86,40 @@ def requirements(self):
             self.requires("fp16/cci.20210320")
 
+    def build_requirements(self):
+        self.tool_requires("cmake/3.24.0")
```
Upstream requires 3.16 https://github.com/tensorflow/tensorflow/blob/1a17caa08363e88204a285f6a8678c3b96e96e0e/tensorflow/lite/CMakeLists.txt#L29

So this is the latest (avoid conflicts else where)

talyz replied:
This was done based on this: https://github.com/conan-io/conan-center-index/pull/11477#discussion_r947110468

prince-chrismc replied:
My fault, accepting full responsibility 😆 

![image](https://user-images.githubusercontent.com/16867443/186461649-c27c07f7-bc4b-402a-a239-802a45ff7d96.png)

talyz replied:
Do you want me to change it to 3.16 instead? It doesn't seem to be in CCI anymore - 3.19.8 is the oldest version listed in the cmake conandata.yml, at least.

uilianries replied:
@talyz nope, it's fine using latest version to avoid conflicts. Just checking it's necessary to install a new cmake version 👍

prince-chrismc commented on recipes/tensorflow-lite/all/conandata.yml:
```diff
@@ -2,11 +2,33 @@ sources:
+      patch_type: "conan"
       base_path: "src"
-    - patch_file: "patches/link_targets.patch"
+    - patch_file: "patches/remove_simple_memory_arena_debug_dump.patch"
```
Please submit a PR upstream 🙏

prince-chrismc commented:
https://github.com/tensorflow/tensorflow/issues/57461
